### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jiangliu**
Refactor the vhost-user-backend framework by:
1) split it into several subsystems
2) enhance documentation
3) upgrade dependency
4) add some unit test cases

This is a preparation for support of guest memory dirty tracking.

### CODE REVIEW

**Review by sboeuf — DISMISSED** [review]
Great refactoring, thanks @jiangliu for doing this :)

File: src/vring.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        if let Some(kick) = &state.kick {
+            kick.read()?;
+        }
+
+        Ok(state.enabled)
+    }
+
+    pub(crate) fn set_call(&self, file: Option<File>) {
+        // SAFETY: see comment in set_vring_kick()
```
Comment (stefano-garzarella) [inline src/vring.rs:122]: `s/set_vring_kick/set_kick`

File: src/vring.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        self.get_mut().call = file.map(|f| unsafe { EventFd::from_raw_fd(f.into_raw_fd()) });
+    }
+
+    pub(crate) fn set_err(&self, file: Option<File>) {
+        // SAFETY: see comment in set_vring_kick()
```
Comment (stefano-garzarella) [inline src/vring.rs:127]: Ditto

File: src/vring.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// Copyright 2021 Alibaba Cloud Computing. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Struct to maintain state information and manipulae vhost-user queues.
```
Comment (stefano-garzarella) [inline src/vring.rs:6]: `s/manipulae/manipulate`

**Review by stefano-garzarella — CHANGES_REQUESTED** [review]
The refactor looks great! Thanks.

Just a couple of simple improvements in the `vring.rs` comments :-)

File: src/vring.rs (hunk 1 above)
Comment (jiangliu) [reply src/vring.rs]: fixed

**Review by stefano-garzarella — APPROVED** [review]

**Review by stefano-garzarella — DISMISSED** [review]

**Review by sboeuf — APPROVED** [review]

File: src/vring.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    /// Get a mutable guard to the underlying raw `VringState` object.
+    pub fn get_mut(&self) -> RwLockWriteGuard<VringState<M>> {
+        self.state.write().unwrap()
+    }
+
+    /// Add an used descriptor into the used queue.
+    pub fn add_used(&self, mem: &M::M, desc_index: u16, len: u32) -> Option<u16> {
+        self.get_mut()
+            .get_queue_mut()
+            .add_used(mem, desc_index, len)
```
Comment (slp) [inline src/vring.rs:69]: The current signature for `virtio_queue::Queue::add_used` is like this:

```
 pub fn add_used(&mut self, head_index: u16, len: u32) -> Result<(), Error>
````

Does this PR require some pending change to `virtio_queue`?

File: src/vring.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    }
+}
+
+/// Struct to maintain state information and manipulate a vhost-user queue.
+#[derive(Clone)]
+pub struct Vring<M: GuestAddressSpace = GuestMemoryAtomic<GuestMemoryMmap>> {
+    state: Arc<RwLock<VringState<M>>>,
+}
+
+impl<M: GuestAddressSpace> Vring<M> {
```
Comment (slp) [inline src/vring.rs:54]: Do you think we should add wrappers for `enable_notification`, `disable_notification` and `needs_notification` too?

File: src/event_loop.rs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    /// it, the backend implementation of `handle_event` will be called. This lets entire control
+    /// to the caller about what needs to be done for this special event, without forcing it
+    /// to run its own dedicated epoll loop for it.
+    pub fn register_listener(
+        &self,
+        fd: RawFd,
+        ev_type: epoll::Events,
+        data: u64,
+    ) -> result::Result<(), io::Error> {
+        if data < self.vrings.len() as u64 {
```
Comment (slp) [inline src/event_loop.rs:129]: Why `data` needs to be lower than the number of vrings? This seems to break `set_vring_kick` here.

**Review by slp — CHANGES_REQUESTED** [review]
Looks really good, thanks a lot for working on this. I've added some comments here and there, but nothing too serious.

File: src/event_loop.rs (hunk 6 above)
Comment (jiangliu) [reply src/event_loop.rs:129]: `data` is used as "event index",  and value lower than self.vrings.len() are reserved for vrings. Please refer to event_loop.rs:214.
```
        if (device_event as usize) < self.vrings.len() {
            let vring = &self.vrings[device_event as usize];
        }
```

File: src/vring.rs (hunk 5 above)
Comment (jiangliu) [reply src/vring.rs:54]: sure, will take care of them.

File: src/vring.rs (hunk 4 above)
Comment (jiangliu) [reply src/vring.rs:69]: yes, I do have some pending PRs for vm-virtio, all our existing projects are using signature of 
```
pub fn add_used(&self, mem: &M::M, desc_index: u16, len: u32) -> Option<u16>
```
I will change the signature to follow current interface of vm-virtio/virtio-queue.

File: src/vring.rs (hunk 5 above)
Comment (jiangliu) [reply src/vring.rs:55]: fixed

File: src/event_loop.rs (hunk 6 above)
Comment (slp) [reply src/event_loop.rs:130]: But, in that case, how do event indexes for the vrings get registered into a `VringEpollHandler`?

Comment (jiangliu) [reply src/event_loop.rs]: Good point, it's also called by handler.rs. Then we need to update the doc.

Comment (jiangliu) [reply src/event_loop.rs]: Fixed

File: src/vring.rs [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    /// Enable event notification for queue.
+    pub fn enable_notification(&self) -> Result<bool, VirtQueError> {
+        self.get_mut().get_queue_mut().enable_notification()
+    }
+
+    /// Disable event notification for queue.
+    pub fn disable_notification(&self) -> Result<bool, VirtQueError> {
+        self.get_mut().get_queue_mut().enable_notification()
```
Comment (slp) [inline src/vring.rs:87]: This should be a call to `disable_notification`

**Review by slp — CHANGES_REQUESTED** [review]
I've been playing with this version of vhost-user-backend a bit, and I had some trouble finding a way to implement an optimal path for the single-threaded use case:

Let's imagine we have something like:

```
    fn process_queue_single(&mut self, vring: &Vring) -> Result<bool> {
        let mut used_any = false;
        let mem = match &self.mem {
            Some(m) => m.memory(),
            None => return Err(Error::NoMemoryConfigured),
        };

        while let Some(avail_desc) = vring
            .get_mut()
            .get_queue_mut()
            .iter()
            .map_err(|_| Error::IterateQueue)?
            .next()
        {
            // Process the request...

            if self.event_idx {
                if vring.add_used(head_index, 0).is_err() {
                    warn!("Couldn't return used descriptors to the ring");
                }

                match vring.needs_notification() {
                    Err(_) => {
                        warn!("Couldn't check if queue needs to be notified");
                        vring.signal_used_queue().unwrap();
                    }
                    Ok(needs_notification) => {
                        if needs_notification {
                            vring.signal_used_queue().unwrap();
                        }
                    }
                }
            } else {
                if vring.add_used(head_index, 0).is_err() {
                    warn!("Couldn't return used descriptors to the ring");
                }
                vring.signal_used_queue().unwrap();
            }
        }

        Ok(used_any)
    }
```

Here we have two problems:

1. It will abort on `vring.add_used()` and/or `vring.signal_used_queue` to prevent a deadlock, as the same thread is trying to acquire `Vring.state` twice.
2. We're acquiring and releasing the lock too many times (wasted CPU cycles).

The best solution I've found is moving `add_used`, `needs_notification` and `signal_used_queue` to `VringState`, turning the version of those methods in `Vring` into proxies for the firsts. That way, you can acquire the lock once by acquiring a reference to `VringState`, and operate the queue entirely from it:

```
    fn process_queue_single(&mut self, vring: &Vring) -> Result<bool> {
        let mut used_any = false;
        let mem = match &self.mem {
            Some(m) => m.memory(),
            None => return Err(Error::NoMemoryConfigured),
        };

        let mut vring_state = vring.get_mut();

        while let Some(avail_desc) = vring_state
            .get_queue_mut()
            .iter()
            .map_err(|_| Error::IterateQueue)?
            .next()
        {
            // Process the request...

            if self.event_idx {
                if vring_state.add_used(head_index, 0).is_err() {
                    warn!("Couldn't return used descriptors to the ring");
                }

                match vring_state.needs_notification() {
                    Err(_) => {
                        warn!("Couldn't check if queue needs to be notified");
                        vring_state.signal_used_queue().unwrap();
                    }
                    Ok(needs_notification) => {
                        if needs_notification {
                            vring_state.signal_used_queue().unwrap();
                        }
                    }
                }
            } else {
                if vring_state.add_used(head_index, 0).is_err() {
                    warn!("Couldn't return used descriptors to the ring");
                }
                vring_state.signal_used_queue().unwrap();
            }
        }

        Ok(used_any)
    }
```

This works for me, but I'm not sure if this goes against the locking philosophy you wanted to implement, or if is there a better pattern for this use case.

File: src/vring.rs (hunk 7 above)
Comment (jiangliu) [reply src/vring.rs:87]: sorry, will fix it:(

**Comment by jiangliu** [comment]
My personal preference is to use VringState directly instead of the Vring adapter.
The mere goal of Vring adapter is to ease migrate existing code base, but all our existing code are using VringState directly.
The proposed solution for single-threaded should work. 

So the only issue left is that a high performance virtio backend driver needs to implement two request handling routines, one for single-thread and one for multi-threading. I think this tradeoff is acceptable because we only need to do this for really high performance virtio devices, such as virtio-blk/scsi/fs etc.

**Comment by slp** [comment]
> So the only issue left is that a high performance virtio backend driver needs to implement two request handling routines, one for single-thread and one for multi-threading. I think this tradeoff is acceptable because we only need to do this for really high performance virtio devices, such as virtio-blk/scsi/fs etc.

I agree.

File: src/vring.rs (hunk 7 above)
Comment (jiangliu) [reply src/vring.rs]: Fixed

**Comment by slp** [comment]
@jiangliu Do you want to implement the single-thread optimizations (moving `add_used`, `needs_notification` and `signal_used_queue` to `VringState`) in this PR or in a future one?

**Comment by jiangliu** [comment]
> @jiangliu Do you want to implement the single-thread optimizations (moving `add_used`, `needs_notification` and `signal_used_queue` to `VringState`) in this PR or in a future one?

I feel it deserve a dedicated MR, we need add some one trait to abstract Vring and VringState, just like QueueState.

**Comment by slp** [comment]
> > @jiangliu Do you want to implement the single-thread optimizations (moving `add_used`, `needs_notification` and `signal_used_queue` to `VringState`) in this PR or in a future one?
> 
> I feel it deserves a dedicated MR:) We need add one trait to abstract Vring and VringState, like what we have done for QueueState.

Okay, then let's get this one merged :-)

**Review by slp — DISMISSED** [review]

**Review by stefano-garzarella — DISMISSED** [review]

**Review by sboeuf — APPROVED** [review]

**Review by slp — APPROVED** [review]

**Pull request merged**